evalml/pipelines/components/transformers/preprocessing/polynomial_decomposer.py

```diff
@@ -0,0 +1,281 @@
+        y_ww = infer_feature_types(y)
+        y_detrended = self._component_obj.transform(y_ww)
+
+        y = y.copy()
+        if not isinstance(y.index, pd.DatetimeIndex):
+            y = self._set_time_index(X, y)
+
+        if isinstance(y.index, pd.DatetimeIndex) and y.index.freqstr != self.frequency:
+            raise ValueError(
+                f"Cannot transform given data with frequency {y.index.freqstr}. "
+                f"Transformer was trained on data with frequency {self.frequency}"
+            )
+
+        if isinstance(y.index, pd.DatetimeIndex):
+            # Repeat the seasonal signal over the target data
+            seasonal = np.tile(
+                self.seasonality.T, len(y_detrended) // self.periodicity + 1
+            ).T[: len(y_detrended)]
+
+        y_t = pd.Series(y_detrended - seasonal).set_axis(y.index)
+        y_t.ww.init(logical_type="double")
+        return X, y_t
+
+    def fit_transform(self, X, y=None):
+        """Removes fitted trend and seasonality from target variable.
+
+        Args:
+            X (pd.DataFrame, optional): Ignored.
+            y (pd.Series): Target variable to detrend and deseasonalize.
+
+        Returns:
+            tuple of pd.DataFrame, pd.Series: The first element are the input features returned without modification.
+                The second element is the target variable y with the fitted trend removed.
+        """
+        return self.fit(X, y).transform(X, y)
+
+    def inverse_transform(self, y):
+        """Adds back fitted trend and seasonality to target variable.
+
+        The polynomial trend is added in the traditional way, calling the detrender's inverse_transform().
+        Then the seasonality is projected forward to determine and re-add the seasonality.
+
+        Args:
+            y (pd.Series): Target variable.
+
+        Returns:
+            tuple of pd.DataFrame, pd.Series: The first element are the input features returned without modification.
+                The second element is the target variable y with the trend and seasonality added back.
+
+        Raises:
+            ValueError: If y is None.
+        """
+        if y is None:
+            raise ValueError("y cannot be None for PolynomialDecomposer!")
+        y_ww = infer_feature_types(y)
+
+        # Add polynomial trend back to signal
+        y_retrended = self._component_obj.inverse_transform(y_ww)
+
+        # Determine where the seasonality starts
+        first_index_diff = y_ww.index[0] - self.seasonality.index[0]
+        # TODO: Write tests to test different time series frequencies.
+        if self.frequency == "D":
+            delta = timedelta(days=1)
+            period = timedelta(days=self.periodicity)
+
+        # Cycle through the saved seasonality to match the first index of the transform data and project forward to the last index
+        transform_first_ind = int((first_index_diff % period) / delta)
```

Comment:
Let's explicitly test this in our unit tests - it would be great to cover all the frequencies and various positions to project forward with. Would be great to keep a more verbose comment here as well to explain the array arithmetic.

Reply:
OK, I'm going to break this into two tests - one with a constant frequency, just testing different indices and another one to test different frequencies.

Reply:
Done!